crypto/asn1/d2i_pr.c

```diff
@@ -40,24 +40,29 @@ d2i_PrivateKey_decoder(int keytype, EVP_PKEY **a, const unsigned char **pp,
         if (key_name == NULL)
             return NULL;
     }
-    if (a != NULL && *a != NULL)
-        ppkey = a;
 
     for (i = 0;  i < (int)OSSL_NELEM(input_structures); ++i) {
         const unsigned char *p = *pp;
 
+        if (a != NULL && (bak_a = *a) != NULL)
+            ppkey = a;
         dctx = OSSL_DECODER_CTX_new_for_pkey(ppkey, "DER",
                                              input_structures[i], key_name,
                                              EVP_PKEY_KEYPAIR, libctx, propq);
+        if (a != NULL)
+            *a = bak_a;
         if (dctx == NULL)
             return NULL;
```

Comment:
There is actually one more bug here - if there is no decoder at all for "type-specific" input structure, this will prematurely exit as `dctx` will be NULL. I am really curious why do we have a cycle over input structures here at all, why don't we just pass NULL straight away and leave it up to the decoding process to select the input structure.

Reply:
> There is actually one more bug here - if there is no decoder at all for "type-specific" input structure, this will prematurely exit as `dctx` will be NULL.

Good point. All right, fixed that as well.

> I am really curious why do we have a cycle over input structures here at all, why don't we just pass NULL straight away and leave it up to the decoding process to select the input structure.

I have no idea since I'm not familiar with that machinery. @slontis should know.
Yet please place this in a separate issue/PR since this is certainly beyond the scope of this PR.

Reply:
Can we finish this PR now?

Reply:
Re input structures, it's to mimic what legacy does.  The "old" EVP_PKEY_ASN1_METHOD decode functions decode type specific structures, while the newer ones decode PKCS8...